xmtp_mls/src/worker.rs

```diff
@@ -0,0 +1,130 @@
+            self.metrics.as_ref().lock().clone().map(|arc| {
+                // This is safe because we verified Metric == SyncMetric
+                unsafe { std::mem::transmute::<Arc<WorkerMetrics<Metric>>, Arc<WorkerMetrics<SyncMetric>>>(arc.clone()) }
+            })
+        } else {
+            None
+        }
+    }
+
+    fn spawn(&self) {
+        (self.spawn_fn)();
+    }
+}
+
+pub struct WorkerRunner<Core, Metric = NoMetric> {
+    pub metrics: Arc<Mutex<Option<Arc<WorkerMetrics<Metric>>>>>,
+    spawn_fn: Box<dyn Fn() + Send + Sync>,
+    _core: PhantomData<Core>,
+}
+
+impl<Core, Metric> WorkerRunner<Core, Metric>
+where
+    Metric: PartialEq + Hash + Send + Sync + 'static,
+{
+    pub fn register_new_worker<ApiClient, Db>(client: &Client<ApiClient, Db>)
+    where
+        ApiClient: XmtpApi + Send + Sync + 'static,
+        Db: XmtpDb + 'static,
+        Core: Worker<ApiClient, Db, Metric> + 'static,
+    {
+        let metrics = Arc::new(Mutex::default());
+        let runner = WorkerRunner {
+            metrics: metrics.clone(),
+            spawn_fn: Box::new({
+                let client = client.clone();
+                let metrics = metrics.clone();
+                move || {
+                    Self::spawn_worker_internal(&client, &metrics);
+                }
+            }),
+            _core: PhantomData::<Core>,
+        };
+
+        runner.spawn();
+        let kind = Core::kind();
+        let runner = Box::new(runner);
+        client.context.workers.lock().insert(kind, runner as Box<_>);
+    }
+
+    pub(crate) fn spawn_worker_internal<ApiClient, Db>(
+        client: &Client<ApiClient, Db>,
+        metrics: &Arc<Mutex<Option<Arc<WorkerMetrics<Metric>>>>>,
+    ) where
+        ApiClient: XmtpApi + 'static,
+        Db: XmtpDb + Send + Sync + 'static,
+        Core: Worker<ApiClient, Db, Metric> + 'static,
+    {
+        let mut core = Core::init(client);
+        *metrics.lock() = core.metrics();
+
+        xmtp_common::spawn(None, async move {
+            loop {
+                if let Err(err) = core.run_tasks().await {
+                    if err.needs_db_reconnect() {
+                        tracing::warn!("Pool disconnected. task will restart on reconnect");
+                        break;
+                    } else {
+                        tracing::error!("Worker error: {err:?}");
+                        xmtp_common::time::sleep(WORKER_RESTART_DELAY).await;
+                        tracing::info!("Restarting sync worker...");
+                    }
+                }
+            }
+        });
+    }
+}
+
+#[async_trait::async_trait]
+pub trait Worker<ApiClient, Db, Metric = NoMetric>
+where
+    Self: Send + Sync,
+    ApiClient: XmtpApi,
+    Db: xmtp_db::XmtpDb,
+    Metric: PartialEq + Hash + Send + Sync + 'static,
+{
+    type Error: NeedsDbReconnect + Debug + Send;
+
+    fn kind() -> WorkerKind;
+    fn init(client: &Client<ApiClient, Db>) -> Self;
```

Comment:
I think it could be useful to refactor a bit here

Why can't the init method accept a &self parameter?

Then in the Worker registration, you could accept a `W: Worker`. Then, kind would also be able to accept self by reference, which would allow this Worker Trait to be Boxable